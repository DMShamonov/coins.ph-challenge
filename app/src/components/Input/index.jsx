// @flow

import * as React from 'react';
import _uniqueId from 'lodash/uniqueId';
import classnames from 'classnames';

import type { PropsType } from './types.js.flow';

export default function Input({
  id,
  type = 'text',
  name,
  label,
  placeholder,
  value,
  defaultValue,
  autoFocus,
  autoComplete = 'on',
  required,
  disabled,
  className,
  onChange,
}: PropsType): React.Node {
  const _id = id || name || _uniqueId();

  return (
    <div className={classnames('input', className)}>
      <label htmlFor={_id}>
        <span className="input__label">{label}</span>
        <input
          id={_id}
          name={name}
          value={value}
          defaultValue={defaultValue}
          type={type}
          placeholder={placeholder}
          autoFocus={autoFocus}
          autoComplete={autoComplete}
          required={required}
          disabled={disabled}
          className="input__field"
          onChange={onChange}
        />
      </label>
    </div>
  );
}
