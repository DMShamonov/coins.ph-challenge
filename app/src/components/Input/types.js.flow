// @flow

export type PropsType = {|
  id?: string,
  type?: (
    | 'text'
    | 'email'
    | 'password'
  ),
  name?: string,
  label?: string,
  placeholder?: string,
  value?: string,
  defaultValue?: string,
  autoFocus?: boolean,
  autoComplete?: (
    | 'off'
    | 'on'
  ),
  required?: boolean,
  disabled?: boolean,
  className?: string,
  onChange?: SyntheticInputEvent<HTMLInputElement> => void,
|}
