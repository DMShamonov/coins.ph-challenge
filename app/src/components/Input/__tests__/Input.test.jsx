import React from 'react';
import { shallow } from 'enzyme';

import Input from '../index';

describe('`Input` component should render correctly', () => {
  test('with default props', () => {
    expect(shallow(<Input />)).toMatchSnapshot();
  });

  test('with `id` prop', () => {
    expect(shallow(<Input id="__ID__" />)).toMatchSnapshot();
  });

  test('when `type` prop is `number`', () => {
    expect(shallow(<Input type="number" />)).toMatchSnapshot();
  });

  test('with `name` prop', () => {
    expect(shallow(<Input name="__NAME__" />)).toMatchSnapshot();
  });

  test('with `label` prop', () => {
    expect(shallow(<Input label="__LABEL__" />)).toMatchSnapshot();
  });

  test('with `placeholder` prop', () => {
    expect(shallow(<Input placeholder="__PLACEHOLDER__" />)).toMatchSnapshot();
  });

  test('with `value` prop', () => {
    expect(shallow(<Input value="__VALUE__" />)).toMatchSnapshot();
  });

  test('with `defaultValue` prop', () => {
    expect(shallow(<Input defaultValue="__DEFAULT_VALUE__" />)).toMatchSnapshot();
  });

  test('when `autoFocus` prop is `true`', () => {
    expect(shallow(<Input autoFocus />)).toMatchSnapshot();
  });

  test('when `autoComplete` prop is `off`', () => {
    expect(shallow(<Input autoComplete="off" />)).toMatchSnapshot();
  });

  test('when `required` prop is `true`', () => {
    expect(shallow(<Input required />)).toMatchSnapshot();
  });

  test('when `disabled` prop is `true`', () => {
    expect(shallow(<Input disabled />)).toMatchSnapshot();
  });

  test('with `className` prop', () => {
    expect(shallow(<Input className="__CLASSNAME__" />)).toMatchSnapshot();
  });

  test('with `onChange` prop', () => {
    expect(shallow(<Input onChange={jest.fn()} />)).toMatchSnapshot();
  });
});

describe('Check `Input` props', () => {
  describe('Check callback props', () => {
    test('`onChange` called', () => {
      const onChange = jest.fn();
      const wrapper = shallow(<Input onChange={onChange} />);

      wrapper.find('input').simulate('change');

      expect(onChange).toHaveBeenCalled();
    });
  });
});
