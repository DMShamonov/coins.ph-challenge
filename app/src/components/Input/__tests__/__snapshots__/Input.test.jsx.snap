// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Input\` component should render correctly when \`autoComplete\` prop is \`off\` 1`] = `
<div
  className="input"
>
  <label
    htmlFor="8"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="off"
      className="input__field"
      id="8"
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly when \`autoFocus\` prop is \`true\` 1`] = `
<div
  className="input"
>
  <label
    htmlFor="7"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      autoFocus={true}
      className="input__field"
      id="7"
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly when \`disabled\` prop is \`true\` 1`] = `
<div
  className="input"
>
  <label
    htmlFor="10"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      disabled={true}
      id="10"
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly when \`required\` prop is \`true\` 1`] = `
<div
  className="input"
>
  <label
    htmlFor="9"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      id="9"
      required={true}
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly when \`type\` prop is \`number\` 1`] = `
<div
  className="input"
>
  <label
    htmlFor="2"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      id="2"
      type="number"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly with \`className\` prop 1`] = `
<div
  className="input __CLASSNAME__"
>
  <label
    htmlFor="11"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      id="11"
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly with \`defaultValue\` prop 1`] = `
<div
  className="input"
>
  <label
    htmlFor="6"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      defaultValue="__DEFAULT_VALUE__"
      id="6"
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly with \`id\` prop 1`] = `
<div
  className="input"
>
  <label
    htmlFor="__ID__"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      id="__ID__"
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly with \`label\` prop 1`] = `
<div
  className="input"
>
  <label
    htmlFor="3"
  >
    <span
      className="input__label"
    >
      __LABEL__
    </span>
    <input
      autoComplete="on"
      className="input__field"
      id="3"
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly with \`name\` prop 1`] = `
<div
  className="input"
>
  <label
    htmlFor="__NAME__"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      id="__NAME__"
      name="__NAME__"
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly with \`onChange\` prop 1`] = `
<div
  className="input"
>
  <label
    htmlFor="12"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      id="12"
      onChange={[MockFunction]}
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly with \`placeholder\` prop 1`] = `
<div
  className="input"
>
  <label
    htmlFor="4"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      id="4"
      placeholder="__PLACEHOLDER__"
      type="text"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly with \`value\` prop 1`] = `
<div
  className="input"
>
  <label
    htmlFor="5"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      id="5"
      type="text"
      value="__VALUE__"
    />
  </label>
</div>
`;

exports[`\`Input\` component should render correctly with default props 1`] = `
<div
  className="input"
>
  <label
    htmlFor="1"
  >
    <span
      className="input__label"
    />
    <input
      autoComplete="on"
      className="input__field"
      id="1"
      type="text"
    />
  </label>
</div>
`;
